mods/cef-titlebar-enabler-universal.wh.cpp

```diff
@@ -0,0 +1,259 @@
+    * 1.1.68-1.1.70: Window control hiding doesn't work yet
+    * 1.1.74: Last version to support proper DWM window controls
+    * Native DWM window controls are invisible since 1.1.75 as it began hooking window messages to support Windows 11 maximize button hover menu
+    * 1.1.85: Last version to support proper non-DWM window controls
+    * They are still visible after 1.1.85, but they don't respond to clicks
+    * 1.2.30: Only x86-64 is supported from this version
+    * 1.2.45: Last version to support disabling the global navbar
+    * Try toggling hardware acceleration multiple times (both in window menu and preferences) if you want a proper window icon. It may work on recent-ish versions
+*/
+// ==/WindhawkModReadme==
+
+// ==WindhawkModSettings==
+/*
+- showframe: true
+  $name: Enable native frames and title bars
+- showmenu: true
+  $name: Show the menu button
+  $description: Disabling this also prevents opening the Spotify menu with the Alt key
+- showcontrols: false
+  $name: Show Spotify's custom window controls
+*/
+// ==/WindhawkModSettings==
+
+#include <libloaderapi.h>
+#include <windhawk_utils.h>
+
+#define CEF_CALLBACK __stdcall
+
+struct cte_settings {
+    BOOL showframe;
+    BOOL showmenu;
+    BOOL showcontrols;
+} cte_settings;
+
+typedef struct cte_offset {
+  int ver_major;
+  int ver_minor;
+  int offset;
+} cte_offset_t;
+
+cte_offset_t is_frameless_offsets[] = {
+    {90, 6, 0x48}, // Spotify 1.1.60 to 1.1.62
+    {91, 1, 0x48}, // Spotify 1.1.63 to 1.1.67
+    {91, 3, 0x50}, // Spotify 1.1.68 to 1.1.70
+    {94, 0, 0x50}, // Spotify 1.1.71
+    {101, 0, 0x50}, // Spotify 1.1.88
+    {102, 0, 0x54}, // Spotify 1.1.89
+    {106, 0, 0x54}, // Spotify 1.1.97
+    {120, 0, 0xc8}, // Spotify 1.2.30
+    {124, 0, 0xd0}, // Spotify 1.2.38
+    {130, 0, 0xd0} // Spotify 1.2.52
+};
+
+cte_offset_t add_child_view_offsets[] = {
+    {90, 6, NULL}, // Spotify 1.1.60 to 1.1.62, not needed (use SpotifyNoControls)
+    {91, 1, NULL}, // Spotify 1.1.63 to 1.1.67, not needed (use SpotifyNoControls)
+    {91, 3, NULL}, // Spotify 1.1.68 to 1.1.70, not working
+    {94, 0, 0xf0}, // Spotify 1.1.71
+    {102, 0, 0xf0}, // Spotify 1.1.89
+    {106, 0, 0xf0}, // Spotify 1.1.97
+    {120, 0, 0x1e0}, // Spotify 1.2.30
+    {124, 0, 0x1e8}, // Spotify 1.2.38
+    {130, 0, 0x1e8} // Spotify 1.2.52
+};
+
+cte_offset_t is_frameless_offset = {0, 0, NULL};
+cte_offset_t add_child_view_offset = {0, 0, NULL};
+
+typedef int CEF_CALLBACK (*is_frameless_t)(struct _cef_window_delegate_t* self, struct _cef_window_t* window);
+int CEF_CALLBACK is_frameless_hook(struct _cef_window_delegate_t* self, struct _cef_window_t* window) {
+    Wh_Log(L"is_frameless_hook");
+    return 0;
+}
+
+typedef _cef_window_t* (*cef_window_create_top_level_t)(void* delegate);
+cef_window_create_top_level_t cef_window_create_top_level_original;
+_cef_window_t* cef_window_create_top_level_hook(void* delegate) {
```

Comment:
The calling convention is missing here, which is likely to cause a crash for 32-bit targets. Please add calling conventions or, if the mod doesn't support 32-bit, limit it to 64-bit, which will also make compilation faster. See:
https://github.com/ramensoftware/windhawk/wiki/Development-tips#hooks-and-calling-conventions

Reply:
Done

Reply:
This mod already worked fine without adding a calling convention on Spotify 1.1.x (which only had 32-bit releases). As adding a calling convention did not cause any harm, I added it anyway.

However, adding it to `cef_panel_create` caused Spotify 1.1.x to crash, so I did not add it there.

Also, as stated in the readme, 32-bit releases of 1.2.x are not supported. I did not figure out how to get it to work on those versions. (This mod simply does nothing on them.)

Reply:
> As adding a calling convention did not cause any harm

That's not enough to make sure that the calling convention is correct. If you don't have headers or documentation, you can check the calling convention via assembly. Usually it's either cdecl or stdcall, but there are other calling conventions too, such as fastcall. If you need help with that, you can upload the binary and I'll take a look.

> However, adding it to `cef_panel_create` caused Spotify 1.1.x to crash

By not adding a calling convention, you're leaving it up to the compiler to pick one. I'm not sure what's the default that Windhawk's compiler uses, maybe cdecl, but I'd rather not rely on it, as it might change in the future. It's best to specify the calling convention explicitly.

> Also, as stated in the readme, 32-bit releases of 1.2.x are not supported.

In this case, how about returning `FALSE` for these versions to have an indication the the mod wasn't loaded successfully? It can be something like:
```cpp
#ifndef _WIN64
  if (...1.2.x version...) {
    return FALSE;
  }
#endif
```

Reply:
Inspected the binary with IDA and found out those two functions are cdecl. Btw there are no calling conventions defined for those functions in the CEF source code. It seems it's just using the compiler default.

Also found out that I simply missed that struct sizes are different per architecture; now it supports both x86 and x86-64 on all supported versions.

I also added support for more versions and tested all Spotify releases since 1.1.60. Now, there are no unsupported versions between CEF 90.4-132 and Spotify 1.1.60-1.2.52.